### PR TITLE
fix(duckdb): use functions for temporal literals

### LINKED ISSUE #7722: bug: CompileError: No literal value renderer is available for literal value "datetime.datetime(1998, 9, 2, 0, 0)" with datatype DATETIME
### What happened?

```python
import ibis

from ibis import _

con = ibis.connect("duckdb://")

# generate tpc-h data
con.raw_sql("call dbgen(sf=1)")

# construct and run q1
from datetime import datetime

var_1 = datetime(1998, 9, 2)
q = con.table("lineitem")
q_final = (
    q.filter(_.l_shipdate <= var_1)
    .group_by(["l_returnflag", "l_linestatus"])
    .agg(
        sum_qty=_.l_quantity.sum(),
        sum_base_price=_.l_extendedprice.sum(),
        sum_disc_price=(_.l_extendedprice * (1 - _.l_discount)).sum(),
        sum_charge=(
            _.l_extendedprice * (1.0 - _.l_discount) * (1.0 + _.l_tax)
        ).sum(),
        avg_qty=_.l_quantity.mean(),
        avg_price=_.l_extendedprice.mean(),
        avg_disc=_.l_discount.mean(),
        count_order=_.count(),

    )
    .order_by(["l_returnflag", "l_linestatus"])
)
q_final
```

this outputs:

```
r0 := DatabaseTable: lineitem
  l_orderkey      int32
  l_partkey       int32
  l_suppkey       int32
  l_linenumber    int32
  l_quantity      decimal(15, 2)
  l_extendedprice decimal(15, 2)
  l_discount      decimal(15, 2)
  l_tax           decimal(15, 2)
  l_returnflag    string
  l_linestatus    string
  l_shipdate      date
  l_commitdate    date
  l_receiptdate   date
  l_shipinstruct  string
  l_shipmode      string
  l_comment       string

r1 := Aggregation[r0]
  metrics:
    sum_qty:        Sum(r0.l_quantity)
    sum_base_price: Sum(r0.l_extendedprice)
    sum_disc_price: Sum(r0.l_extendedprice * 1 - r0.l_discount)
    sum_charge:     Sum(r0.l_extendedprice * 1.0 - r0.l_discount * r0.l_tax + 1.0)
    avg_qty:        Mean(r0.l_quantity)
    avg_price:      Mean(r0.l_extendedprice)
    avg_disc:       Mean(r0.l_discount)
    count_order:    CountStar(r0)
  by:
    l_returnflag: r0.l_returnflag
    l_linestatus: r0.l_linestatus
  predicates:
    r0.l_shipdate <= datetime.datetime(1998, 9, 2, 0, 0)

Selection[r1]
  sort_keys:
    asc r1.l_returnflag
    asc r1.l_linestatus
```

then turning on Ibis interactive mode and printing out the dataframe:

```python
ibis.options.interactive = True

q_final
```

```
---------------------------------------------------------------------------
CompileError                              Traceback (most recent call last)
File [~/repos/ibis/venv/lib/python3.11/site-packages/IPython/core/formatters.py:974](https://file+.vscode-resource.vscode-cdn.net/Users/cody/repos/ibis/~/repos/ibis/venv/lib/python3.11/site-packages/IPython/core/formatters.py:974), in MimeBundleFormatter.__call__(self, obj, include, exclude)
    [971](https://file+.vscode-resource.vscode-cdn.net/Users/cody/repos/ibis/~/repos/ibis/venv/lib/python3.11/site-packages/IPython/core/formatters.py:971)     method = get_real_method(obj, self.print_method)
    [973](https://file+.vscode-resource.vscode-cdn.net/Users/cody/repos/ibis/~/repos/ibis/venv/lib/python3.11/site-packages/IPython/core/formatters.py:973)     if method is not None:
--> [974](https://file+.vscode-resource.vscode-cdn.net/Users/cody/repos/ibis/~/repos/ibis/venv/lib/python3.11/site-packages/IPython/core/formatters.py:974)         return method(include=include, exclude=exclude)
    [975](https://file+.vscode-resource.vscode-cdn.net/Users/cody/repos/ibis/~/repos/ibis/venv/lib/python3.11/site-packages/IPython/core/formatters.py:975)     return None
    [976](https://file+.vscode-resource.vscode-cdn.net/Users/cody/repos/ibis/~/repos/ibis/venv/lib/python3.11/site-packages/IPython/core/formatters.py:976) else:

File [~/repos/ibis/ibis/expr/types/core.py:36](https://file+.vscode-resource.vscode-cdn.net/Users/cody/repos/ibis/~/repos/ibis/ibis/expr/types/core.py:36), in _FixedTextJupyterMixin._repr_mimebundle_(self, *args, **kwargs)
     [35](https://file+.vscode-resource.vscode-cdn.net/Users/cody/repos/ibis/~/repos/ibis/ibis/expr/types/core.py:35) def _repr_mimebundle_(self, *args, **kwargs):
---> [36](https://file+.vscode-resource.vscode-cdn.net/Users/cody/repos/ibis/~/repos/ibis/ibis/expr/types/core.py:36)     bundle = super()._repr_mimebundle_(*args, **kwargs)
     [37](https://file+.vscode-resource.vscode-cdn.net/Users/cody/repos/ibis/~/repos/ibis/ibis/expr/types/core.py:37)     bundle["text/plain"] = bundle["text/plain"].rstrip()
     [38](https://file+.vscode-resource.vscode-cdn.net/Users/cody/repos/ibis/~/repos/ibis/ibis/expr/types/core.py:38)     return bundle

File [~/repos/ibis/venv/lib/python3.11/site-packages/rich/jupyter.py:48](https://file+.vscode-resource.vscode-cdn.net/Users/cody/repos/ibis/~/repos/ibis/venv/lib/python3.11/site-packages/rich/jupyter.py:48), in JupyterMixin._repr_mimebundle_(self, include, exclude, **kwargs)
     [41](https://file+.vscode-resource.vscode-cdn.net/Users/cody/repos/ibis/~/repos/ibis/venv/lib/python3.11/site-packages/rich/jupyter.py:41) def _repr_mimebundle_(
     [42](https://file+.vscode-resource.vscode-cdn.net/Users/cody/repos/ibis/~/repos/ibis/venv/lib/python3.11/site-packages/rich/jupyter.py:42)     self: "ConsoleRenderable",
     [43](https://file+.vscode-resource.vscode-cdn.net/Users/cody/repos/ibis/~/repos/ibis/venv/lib/python3.11/site-packages/rich/jupyter.py:43)     include: Sequence[str],
     [44](https://file+.vscode-resource.vscode-cdn.net/Users/cody/repos/ibis/~/repos/ibis/venv/lib/python3.11/site-packages/rich/jupyter.py:44)     exclude: Sequence[str],
     [45](https://file+.vscode-resource.vscode-cdn.net/Users/cody/repos/ibis/~/repos/ibis/venv/lib/python3.11/site-packages/rich/jupyter.py:45)     **kwargs: Any,
     [46](https://file+.vscode-resource.vscode-cdn.net/Users/cody/repos/ibis/~/repos/ibis/venv/lib/python3.11/site-packages/rich/jupyter.py:46) ) -> Dict[str, str]:
     [47](https://file+.vscode-resource.vscode-cdn.net/Users/cody/repos/ibis/~/repos/ibis/venv/lib/python3.11/site-packages/rich/jupyter.py:47)     console = get_console()
---> [48](https://file+.vscode-resource.vscode-cdn.net/Users/cody/repos/ibis/~/repos/ibis/venv/lib/python3.11/site-packages/rich/jupyter.py:48)     segments = list(console.render(self, console.options))
     [49](https://file+.vscode-resource.vscode-cdn.net/Users/cody/repos/ibis/~/repos/ibis/venv/lib/python3.11/site-packages/rich/jupyter.py:49)     html = _render_segments(segments)
...
   [2678](https://file+.vscode-resource.vscode-cdn.net/Users/cody/repos/ibis/~/repos/ibis/venv/lib/python3.11/site-packages/sqlalchemy/sql/compiler.py:2678)         '"%s" with datatype %s'
   [2679](https://file+.vscode-resource.vscode-cdn.net/Users/cody/repos/ibis/~/repos/ibis/venv/lib/python3.11/site-packages/sqlalchemy/sql/compiler.py:2679)         % (sql_util._repr_single_value(value), type_)
   [2680](https://file+.vscode-resource.vscode-cdn.net/Users/cody/repos/ibis/~/repos/ibis/venv/lib/python3.11/site-packages/sqlalchemy/sql/compiler.py:2680)     )

CompileError: No literal value renderer is available for literal value "datetime.datetime(1998, 9, 2, 0, 0)" with datatype DATETIME
```

### What version of ibis are you using?

master

### What backend(s) are you using, if any?

DuckDB

### Relevant log output

```sh
see above
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by cpcloud**
Fixes #7722.

### CODE REVIEW

**Comment by gforsyth** [comment]
Some snapshots need updating, and somehow Clickhouse has no `_sqlglot_dialect` attribute?

**Comment by cpcloud** [comment]
This is now officially a hellscape

**Comment by gforsyth** [comment]
> This is now officially a hellscape

Should we have an issue label for that?  :fire: :-1: :fire: :sob: :fire: ?

**Comment by cpcloud** [comment]
Let me squash this abomination.

File: ibis/backends/oracle/registry.py [hunk 1]
```diff
@@ -43,13 +43,29 @@ def _corr(t, op):
...
-        and (value := op.value) is not None
+        vparse(sa.__version__) < vparse("2") and dtype.is_uuid()
     ):
         return sa.literal(str(value), type_=t.get_sqla_type(dtype))
+    elif dtype.is_timestamp():
+        if dtype.timezone is not None:
+            return sa.func.to_utc_timestamp_tz(value.isoformat(timespec="microseconds"))
+        return sa.func.to_timestamp(
+            # comma for sep here because T is a special character in Oracle
+            # the FX prefix means "requires an exact match"
```
Comment (gforsyth) [inline ibis/backends/oracle/registry.py:61]: OOF

**Comment by cpcloud** [comment]
Squashed

**Review by gforsyth — APPROVED** [review]

**Pull request merged**